cfpb/consumerfinance.gov
# Add optional SSO support and configuration

This change implements support for SAML2 single sign-on based on [djangosaml2](https://djangosaml2.readthedocs.io/) and [pySAML2](https://pysaml2.readthedocs.io/).

It adds environment variables that, when populated, will enable SAML2 authentication, and the required URL patterns, against the SAML2 identity provider with the given metadata URL using the given entity ID for the Django site. ***Without those environment variables, SAML2 auth will not be enabled and all existing authentication will continue to work as it does now.***

It does require the `xmlsec1` binary to be in place at `/usr/bin/xmlsec1`. That requirement will be handled in our deployment playbooks. There is a corresponding PR for that change and with dev2 configuration in that internal repository.

## How to test this PR

This PR is deployed with SSO enabled to dev2 at present for testing. However, you will need test credentials (I believe everyone on the team should now have those).

This is not easily testable locally. Please reach out to me and I can walk you through it. 

In short:

- Change docker-compose.yml to expose the `python` container on port 443:

   ```yaml
   python:
       …
       ports:
         - 127.0.0.1:443:8000
   ```
- Change Dockerfile run `runsslserver` instead of `runserver`

   ```dockerfile
   CMD ["python", "./cfgov/manage.py", "runsslserver", "0.0.0.0:8000"]
   ```

- Rebuild your `python` Docker container:

   ```shell
   docker-compose build python
   ```

- Add the following variables to your `.env` file (reach out to me for their expected values) and restart the Docker container:

   ```shell
   export SAML_AUTH=True
   export SAML_ROOT_URL="https://<dev2 FQDN>"
   export SAML_ENTITY_ID="<dev2 entity ID>"
   export SAML_METADATA_URL="<dev2 metadata URL>"
   ```

- Edit your `/etc/hosts` to point our dev2 FQDN to 127.0.0.1
- Start or restart your `python` Docker container. Verify that you can access that dev2 FQDN over https locally.
- Try to access the admin at `/admin`. This should redirect you to our test SAML2 provided. As will using dev2 directly, you'll need test credentials to login.

## Checklist
